bin/grade_item.py

```diff
@@ -0,0 +1,412 @@
+    # remove the compilation directory
+    shutil.rmtree(tmp_compilation)
+
+    # copy output files to tmp_work directory
+    copy_contents_into(test_output_path,tmp_work)
+
+    subprocess.call("ls -la "+tmp_work+" >> "+tmp_logs+"/overall.txt",shell=True)
+        
+    # copy validator.out to the current directory
+    shutil.copy (os.path.join(bin_path,obj["gradeable"],"validate.out"),os.path.join(tmp_work,"my_validator.out"))
+
+    # give the untrusted user read/write/execute permissions on the tmp directory & files
+    add_permissions_recursive(tmp_work,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH | stat.S_IWOTH | stat.S_IXOTH,
+                              stat.S_IROTH)
+
+    add_permissions(os.path.join(tmp_work,"my_validator.out"),stat.S_IROTH | stat.S_IXOTH)
+
+    # validator the validator.out as the untrusted user
+    with open(os.path.join(tmp_logs,"validator_log.txt"), 'w') as logfile:
+        validator_success = subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                                             args.which_untrusted,
+                                             os.path.join(tmp_work,"my_validator.out"),
+                                             obj["gradeable"],
+                                             obj["who"],
+                                             str(obj["version"]),
+                                             submission_time],
+                                            stdout=logfile)
+
+    if validator_success == 0:
+        print ("NEW VALIDATOR OK")
+    else:
+        print ("NEW VALIDATOR FAILURE")
+
+    # give permissions to all created files to the hwcron user
+    subprocess.call([os.path.join(SUBMITTY_INSTALL_DIR,"bin","untrusted_execute"),
+                     args.which_untrusted,
+                     "/usr/bin/find",
+                     tmp_work,
+                     "-user",
+                     args.which_untrusted,
+                     "-exec",
+                     "/bin/chmod",
+                     "o+r",
+                     "{}",
+                     ";"])
+
+    # --------------------------------------------------------------------
+    # MAKE RESULTS DIRECTORY & COPY ALL THE FILES THERE
+
+    with open(os.path.join(tmp_logs,"overall.txt"),'a') as f:
+        print ("====================================\nARCHIVING STARTS", file=f)
+
+    subprocess.call("ls -la "+tmp_work+" >> "+tmp_logs+"/overall.txt",shell=True)
+        
+    os.chdir(bin_path)
+
+    # save the old results path!
+    if os.path.isdir(os.path.join(results_path,"OLD")):
+        shutil.move(os.path.join(results_path,"OLD"),
+                    os.path.join(tmp,"OLD_RESULTS"))
+
+    # clean out all of the old files if this is a re-run
+    shutil.rmtree(results_path,ignore_errors=True)
+
+    # Make directory structure in results if it doesn't exist
+    os.makedirs(results_path)
```

Comment:
the comment `if it doesn't exist` makes me feel like you want `exist_ok=True` flag inside `os.makedirs` as otherwise this function will error if the directory exists.